tests/core/test_validation.py

```diff
@@ -983,3 +1095,256 @@ def test_array_from_vtkmatrix(cls, shape):
+                return_original=False,
+            )
+    else:
+        raise RuntimeError("Unexpected test case.")
+
+    # Test abstract wrapper
+    wrapped_abstract = _ArrayLikeWrapper(array_before_wrap)
+    assert np.array_equal(wrapped_abstract._array, expected.array)
+    assert wrapped_abstract.shape == expected.shape
+    assert wrapped_abstract.dtype == expected.dtype
+    assert wrapped_abstract.ndim == expected.ndim
+    assert wrapped_abstract.size == expected.size
+    assert type(wrapped_abstract) is expected.wrapper
+
+    # Test child wrapper
+    wrapped_child = expected.wrapper(array_before_wrap)
+    assert np.array_equal(wrapped_child._array, expected.array)
+    assert wrapped_child.shape == expected.shape
+    assert wrapped_child.dtype == expected.dtype
+    assert wrapped_child.ndim == expected.ndim
+    assert wrapped_abstract.size == expected.size
+    assert type(wrapped_child) is expected.wrapper
+
+    # Test wrapping self returns self
+    wrapped_wrapped = expected.wrapper(wrapped_child)
+    assert wrapped_wrapped is wrapped_child
+
+    assert np.array_equal(wrapped_wrapped._array, expected.array)
+    assert wrapped_wrapped.shape == expected.shape
+    assert wrapped_wrapped.dtype == expected.dtype
+    assert wrapped_wrapped.ndim == expected.ndim
+    assert wrapped_abstract.size == expected.size
+    assert type(wrapped_wrapped) is expected.wrapper
+
+
+ragged_arrays = (
+    [[1, 2, 3], [4, 5], [6, 7, 8, 9]],
+    [np.array([1, 2, 3]), np.array([4, 5]), np.array([6, 7, 8, 9])],
+)
+
+
+@pytest.mark.skipif(
+    sys.platform == 'linux' and sys.version_info < (3, 9, 0),
+    reason="Numpy raise a warning, not an error.",
+)
+@pytest.mark.parametrize('ragged_array', ragged_arrays)
+def test_array_wrapper_ragged_array(ragged_array):
+    # assert casting directly to numpy array raises error
+    with pytest.raises(ValueError, match='inhomogeneous shape'):
+        np.array(ragged_array)
+
+    msg = 'The following array is not valid:\n\t['
+    with pytest.raises(ValueError, match=escape(msg)):
+        _ArrayLikeWrapper(ragged_array)
+
+
+PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(__file__)))
+REVEAL_TYPE_CODE_FILE = PROJECT_ROOT + '/tests/core/test_validation_reveal_type.txt'
+
+
+def reveal_type_from_code(code_snippet):
+    # Call `mypy -c CODE` from the project root dir
+    # This ensures the config is loaded and imports are found
+    # NOTE: running mypy can be slow, avoid making excessive calls
+    cur = os.getcwd()
+    try:
+        os.chdir(PROJECT_ROOT)
+
+        result = mypy_api.run(['-c', code_snippet])
+        assert 'usage: mypy' not in result[1]
+        assert 'Cannot find implementation' not in result[0]
+        stdout = str(result[0]).replace('Tuple', 'tuple')
+        match = findall(r'note: Revealed type is "([^"]+)"', stdout)
+        assert match is not None
+        return match
+
+    finally:
+        os.chdir(cur)
```

Comment:
Running mypy tests like this is good and allows full flexibility and customization over how we run our tests.

One downside though is that it's somewhat non intuitive, and requires creating a separate script file for analysis.

An alternative option is to use `pytest-mypy-plugins` package. This integrates directly with pytest, and is what's used to test the `npt-promote` package.

But, I tried to use it initially and found two main hurdles: (1) specifying the mypy config and path was a bit challenging, and (2) it's not easy to isolate the tests to specific files. So, all 144+ mypy errors that are currently unresolved from #5563 get dumped into each test case and need to be filtered.

Long term though, I think using `pytest-mypy-plugins` is a more elegant solution rather than calling `mypy.api.run`

Reply:
The typing tests have been updated extensively. I think the current test suite works great, no need to use `pytest-mypy-plugins`